nixos/tests/apparmor/default.nix

```
@@ -0,0 +1,133 @@
+        # 2. Sorting according to those strings.
+        # 3. Removing those prepended strings.
+        # 4. Using `diff` against the expected output.
+        with subtest("apparmorRulesFromClosure"):
+            machine.succeed(
+                "${getExe' pkgs.diffutils "diff"} -u ${
+                  pkgs.writeText "expected.rules" (import ./makeExpectedPolicies.nix { inherit pkgs; })
+                } ${
+                  pkgs.runCommand "actual.rules" { preferLocalBuild = true; } ''
+                    ${getExe pkgs.gnused} -e 's:^[^ ]* ${builtins.storeDir}/[^,/-]*-\([^/,]*\):\1 \0:' ${
+                      pkgs.apparmorRulesFromClosure {
+                        name = "ping";
+                        additionalRules = [ "x $path/foo/**" ];
+                      } [ pkgs.libcap ]
+                    } |
+                    ${getExe' pkgs.coreutils "sort"} -n -k1 |
+                    ${getExe pkgs.gnused} -e 's:^[^ ]* ::' >$out
+                  ''
+                }"
+            )
+
+        # Test apparmor profile states by using `diff` against `aa-status`
+        with subtest("apparmorProfileStates"):
+            machine.succeed("${getExe' pkgs.diffutils "diff"} -u <(${getExe' pkgs.apparmor-bin-utils "aa-status"}) ${pkgs.writeText "expected.states" ''
+              apparmor module is loaded.
+              3 profiles are loaded.
+              2 profiles are in enforce mode.
+                 hello
+                 hexdump
+              1 profiles are in complain mode.
+                 sl
+              0 profiles are in prompt mode.
+              0 profiles are in kill mode.
+              0 profiles are in unconfined mode.
+              0 processes have profiles defined.
+              0 processes are in enforce mode.
+              0 processes are in complain mode.
+              0 processes are in prompt mode.
+              0 processes are in kill mode.
+              0 processes are unconfined but have a profile defined.
+              0 processes are in mixed mode.
+            ''}")
+
+        # Test apparmor profile files in /etc/apparmor.d/<name> to be either a correct symlink (sl) or have the right file contents (hello)
+        with subtest("apparmorProfileTargets"):
+            machine.succeed("${getExe' pkgs.diffutils "diff"} -u <(${getExe pkgs.file} /etc/static/apparmor.d/sl) ${pkgs.writeText "expected.link" ''
+              /etc/static/apparmor.d/sl: symbolic link to ${./sl_profile}
+            ''}")
+            machine.succeed("${getExe' pkgs.diffutils "diff"} -u /etc/static/apparmor.d/hello ${pkgs.writeText "expected.content" helloProfileContents}")
+
+
+        with subtest("apparmorProfileEnforce"):
+            machine.succeed("${getExe pkgs.hello} 1> /tmp/test-file")
+            machine.fail("${lib.getExe' pkgs.util-linux "hexdump"} /tmp/test-file") # no access to /etc/os-release granted by apparmor
```

```suggestion
            machine.fail("${lib.getExe' pkgs.util-linux "hexdump"} /tmp/test-file") # no access to /tmp/test-file granted by apparmor
```
unless I missed something

---

Oops, good catch! This is what midnight coding does to a brain, forgot to update the comment!